File: src/main/java/com/github/kindrat/cassandra/client/ui/widget/tableedit/ViewDataConverter.java
package com.github.kindrat.cassandra.client.ui.widget.tableedit;

import com.datastax.driver.core.DataType;
import com.github.kindrat.cassandra.client.util.CqlUtil;
import org.apache.commons.lang3.StringUtils;

import java.util.*;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static java.lang.String.format;
import static java.util.Arrays.stream;

class ViewDataConverter {
    private static final Pattern TABLE_PATTERN;
    private static final Pattern ROW_PATTERN;
    private static final Pattern PRIMARY_KEY_PATTERN;

    static {
        TABLE_PATTERN = Pattern.compile("(?i)(create table)(\\s+)(.*)(\\s+)(\\()([.\\s\\S]*)(\\))");

        List<String> types = stream(DataType.Name.values()).map(Enum::toString).collect(Collectors.toList());
        String combinedTypePattern = String.join("|", types);
        ROW_PATTERN = Pattern.compile(format("(?i)([\"\\-0-9a-z]+)(\\s+)(%s)(\\s*)", combinedTypePattern));

        PRIMARY_KEY_PATTERN = Pattern.compile("(?i)(primary key \\()(.*)(\\))");
    }

    Map.Entry<String, Collection<TableRowEntry>> fromText(String text) {
        Matcher matcher = TABLE_PATTERN.matcher(text);
        boolean tableValid = matcher.find();
        if (tableValid) {
            String tableName = matcher.group(3);
            String rows = matcher.group(6);
            String[] separateRawRows = stream(rows.split("\n"))
                    .map(String::trim)
                    .map(str -> StringUtils.removeEnd(str, ","))
                    .toArray(String[]::new);

            Map<String, TableRowEntry> tableRowEntries = stream(separateRawRows)
                    .filter(row -> ROW_PATTERN.matcher(row).find())
                    .map(this::parseRow)
                    .collect(Collectors.toMap(TableRowEntry::getName, Function.identity()));

            stream(separateRawRows)
                    .filter(row -> PRIMARY_KEY_PATTERN.matcher(row).find())
                    .findAny()
                    .ifPresent(row -> {
                        Matcher pkMatcher = PRIMARY_KEY_PATTERN.matcher(row);
                        boolean isPkLine = pkMatcher.find();
                        if (isPkLine) {
                            String nameLine = pkMatcher.group(2);
                            String[] names = nameLine.split(",");
                            String partitionKey = names[0].trim();
                            tableRowEntries.get(partitionKey).setIsPartitionKey(true);
                            if (names.length > 1) {
                                for (int i = 1; i < names.length; i++) {
                                    String clusteringKey = names[i].trim();
                                    tableRowEntries.get(clusteringKey).setIsClusteringKey(true);
                                }
                            }
                        }
                    });

            return new AbstractMap.SimpleEntry<>(tableName, tableRowEntries.values());
        }
        return new AbstractMap.SimpleEntry<>("", Collections.emptyList());
    }

    private TableRowEntry parseRow(String rawRow) {
        Matcher matcher = ROW_PATTERN.matcher(rawRow);
        boolean matches = matcher.find();
        if (matches) {
            String name = matcher.group(1);
            String type = matcher.group(3);
            DataType.Name dataType = DataType.Name.valueOf(type.toUpperCase());
            return new TableRowEntry(name, dataType, false, false, false);
        } else {
            throw new IllegalStateException("Regex match should happens before");
        }
    }

    String toText(List<TableRowEntry> rows, String table) {
        List<String> rowDefinitions = rows.stream()
                .map(tableRowEntry -> format("%s %s", tableRowEntry.getName(), tableRowEntry.getType()))
                .collect(Collectors.toList());

        rows.stream()
                .filter(entry -> entry.isClusteringKey() || entry.isPartitionKey())
                .sorted(Comparator.comparing(TableRowEntry::isPartitionKey).reversed())
                .map(TableRowEntry::getName)
                .reduce((s, s2) -> s + ", " + s2)
                .ifPresent(s -> rowDefinitions.add(String.format("PRIMARY KEY (%s)", s)));

        String tableRows = rowDefinitions.stream().reduce((s, s2) -> s + ", " + s2).orElse("");
        String tableRawDDL = String.format("CREATE TABLE %s (%s)", table, tableRows);
        return CqlUtil.formatDDL(tableRawDDL);
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/widget/tableedit/ViewDataConverter.kt
package com.github.kindrat.cassandra.client.ui.widget.tableedit

import com.github.kindrat.cassandra.client.util.CqlUtil.PRIMITIVE_TYPES
import com.github.kindrat.cassandra.client.util.CqlUtil.formatDDL
import com.github.kindrat.cassandra.client.util.CqlUtil.parse
import org.apache.commons.lang3.StringUtils
import java.util.*
import java.util.regex.Pattern

object ViewDataConverter {
    private val TABLE_PATTERN: Pattern = Pattern.compile("(?i)(create table)(\\s+)(.*)(\\s+)(\\()([.\\s\\S]*)(\\))")
    private val ROW_PATTERN: Pattern
    private val PRIMARY_KEY_PATTERN: Pattern = Pattern.compile("(?i)(primary key \\()(.*)(\\))")

    init {
        val types: List<String> = PRIMITIVE_TYPES.map{ it.toString() }.toList()
        val combinedTypePattern: String = types.joinToString { "|" }
        ROW_PATTERN = Pattern.compile("(?i)([\"\\-0-9a-z]+)(\\s+)($combinedTypePattern)(\\s*)")
    }

    fun fromText(text: String): Map.Entry<String, Collection<TableRowEntry>> {
        val matcher = TABLE_PATTERN.matcher(text)
        val tableValid = matcher.find()
        if (tableValid) {
            val tableName = matcher.group(3)
            val rows = matcher.group(6)
            val separateRawRows = rows.split("\n")
                    .map { obj: String -> obj.trim { it <= ' ' } }
                    .map { StringUtils.removeEnd(it, ",") }
                    .toList()

            val tableRowEntries: Map<String, TableRowEntry> = separateRawRows
                    .filter { ROW_PATTERN.matcher(it).find() }
                    .map { parseRow(it) }
                    .map { it.name to it }
                    .toMap()

            separateRawRows.find { PRIMARY_KEY_PATTERN.matcher(it).find() }
                    ?.let {
                        val pkMatcher = PRIMARY_KEY_PATTERN.matcher(it)
                        val isPkLine = pkMatcher.find()
                        if (isPkLine) {
                            val nameLine = pkMatcher.group(2)
                            val names = nameLine.split(",").toTypedArray()
                            val partitionKey = names[0].trim()
                            tableRowEntries[partitionKey]?.isPartitionKey = true
                            if (names.size > 1) {
                                for (i in 1 until names.size) {
                                    val clusteringKey = names[i].trim()
                                    tableRowEntries[clusteringKey]?.isClusteringKey = true
                                }
                            }
                        }
                    }
            return AbstractMap.SimpleEntry(tableName, tableRowEntries.values)
        }
        return AbstractMap.SimpleEntry<String, Collection<TableRowEntry>>("", listOf())
    }

    private fun parseRow(rawRow: String): TableRowEntry {
        val matcher = ROW_PATTERN.matcher(rawRow)
        val matches = matcher.find()
        return if (matches) {
            val name = matcher.group(1)
            val type = matcher.group(3)
            TableRowEntry(name, parse(type), isPartitionKey = false, isClusteringKey = false, hasIndex = false)
        } else {
            throw IllegalStateException("Regex match should happens before")
        }
    }

    fun toText(rows: List<TableRowEntry>, table: String): String {
        val rowDefinitions = rows.map { "${it.name} ${it.type}" }.toMutableList()
        rows.filter { it.isClusteringKey || it.isPartitionKey }
                .sortedWith(Comparator.comparing(TableRowEntry::isPartitionKey).reversed())
                .map { it.name }
                .reduce { s: String, s2: String -> "$s, $s2" }
                .apply { rowDefinitions.add("PRIMARY KEY ($this)") }

        val tableRows = rowDefinitions.reduce { s: String, s2: String -> "$s, $s2" }
        return formatDDL("CREATE TABLE $table ($tableRows)")
    }
}